### PR TITLE
WIP: Inplace is now false by default

### PR DESCRIPTION
**Pull request by samysspace**

### CODE REVIEW

**Comment by samysspace** [comment]
If we set the inplace default value to False by default, we run into issues because the way featurize_dataframe() works at the current moment is that it modifies the Pandas table being passed in. However, when inplace is False, we create a whole new DataFrame and return it. However, the tests expect you to have modified the passed in dataframe instead of returning a new one. Due to the way frames and pointers work in Python, it's impossible to change the original variable (the one that is passed into the function) from inside the function if we are creating a whole new object and pointing our parameter to it. I fixed this by changing the tests and setting the dataframe that we pass into the function equal to the result of featurize_dataframe(), e.g. data = multi_f.featurize_dataframe(data, 'x') instead of just multi_f.featurize_dataframe(data, 'x') as we had before. This is why we were getting a bunch of KeyErrors and mismatched length values.

If you want to default to inPlace = False, it must be made clear to the user that you need to **set the dataframe equal to the result of the featurize_dataframe() function**. This is simply optional for the inPlace = True case since we are modifying in place and therefore no pointers get changed. I don't know if this is mergeable since it's a pretty easy fix, but people might have become accustomed to not reassigning their dataframes.

I am still failing two tests but these are because of some Pandas Dataframes issues with the tolist() function but those should be easy to fix. All of the tests that previously had KeyErrors/mismatched column length issues now pass.

**Comment by ardunn** [comment]
We could just trash what we have now and make the "inplace" option the same as how pandas handles it. If inplace, do it inplace and return None. If not inplace, return the updated df. 

So for inplace=False
```
df = some_df
df_featurized = some_featurizer.featurize_dataframe(df, "structure", inplace=False)
# df is still the original some_df and df_featurized is a featurized copy
```

And for inplace=True
```
df = some_df
df_featurized = some_featurizer.featurize_dataframe(df, "structure", inplace=True)
# df is the featurized dataframe, df_featurized is None
```

**Comment by ardunn** [comment]
Also re: reassigning their dataframes

Idk about anyone else, but I always return the df, even when I had inplace=True
```
df = ft.featurize_dataframe(df, "comp")
```

Never just:
```
 ft.featurize_dataframe(df, "comp")
```

I think most of the examples in the notebook are similar to this. So I wouldn't worry too much about it

Just make sure nothing funky is going on with the index!!

**Comment by computron** [comment]
To me it’s clearer if inplace = T does *not* return the data frame. I think this is what pandas does by default. But not a big deal either way IMO.

**Comment by ardunn** [comment]
@computron I agree this should be the behavior (See first comment). Samy any update on finishing this PR? It appears only one test is failing right now so seems pretty close.

Also @samysspace if there is not already a test for inplace=True, could you add it?

**Comment by samysspace** [comment]
Currently working on trying to get that last error fixed. I'll also add in a test for in place = True

**Comment by ardunn** [comment]
Closes #368

File: matminer/featurizers/base.py [hunk 1]
```diff
@@ -260,13 +260,13 @@ def featurize_dataframe(self, df, col_id, ignore_errors=False,
...
-            for k in labels:
-                df[k] = res[k]
+            df[labels] = res[labels]
             return df
         else:
             # Create new dataframe and ensure columns are ordered properly
             new = pd.concat([df, res], axis=1)
-            return new[df.columns.tolist() + res.columns.tolist()]
+            df = new[df.columns.tolist() + res.columns.tolist()]
+            return new
```
Comment (ardunn) [inline matminer/featurizers/base.py]: seems like `df` should be returned, not `new`?

File: matminer/featurizers/tests/test_conversions.py [hunk 2]
```diff
@@ -108,18 +108,27 @@ def test_structure_to_oxidstructure(self):
...
 
         # original is preserved
+        print(type(df["structure"].tolist()[0][0]))
         self.assertEqual(df["structure"].tolist()[0][0].specie, Element("Cl"))
 
         # test in-place
         sto = StructureToOxidStructure(target_col_id=None, overwrite_data=True)
         df = sto.featurize_dataframe(df, 'structure')
-        self.assertEqual(df["structure"].tolist()[0][0].specie.oxi_state, -1)
+        print(type(df["structure"]))
```
Comment (ardunn) [inline matminer/featurizers/tests/test_conversions.py]: in final pr please rm these print statements from the tests

**Comment by ardunn** [comment]
@samysspace some things that might help here:

1. you can run the failing test locally in pycharm without having to run the entire test suite or use the CI. This is much faster and more efficient that pushing every commit to the CI to see if it passes, if the errors are not specific to the CI. --  To do this in Pycharm, go to the line for the test method (e.g., `def test_something(...)`), right click on it, and do "run unittest" - it should hopefully only run that method 

2. Not super important, but try to use somewhat informative commit msgs. This way if someone is trying to read through the commits you made, they can generally get a sense of what each commit is doing (even if it is very short), whereas "test" and "try" as commit messages are difficult to decode what ur doing at a glance

**Comment by samysspace** [comment]
@ardunn Sorry about that, I couldn't figure out a way to run the unit tests locally so every time I made a small change I had to test after committing/pushing (which is why some of these commit messages are so short, usually because I'm just trying to test something that I will change back right after).

**Comment by ardunn** [comment]
No problem, but can you run individual tests locally now? Your commits should be as minimal as possible (i.e., without changing something and changing it right back, otherwise it is hard to follow). 


BTW to run the whole test suite locally, do `python setup.py test` from within your environment in a terminal. For running individual tests I usually use Pycharm. For more info see the pycharm docs!

https://www.jetbrains.com/help/pycharm/performing-tests.html

**Comment by samysspace** [comment]
Yep, sounds good (sorry for the extensive commit history).
